ad/views.py

```diff
@@ -65,4 +66,16 @@ def get_form_class(self):
         return self.form_mapping.get(adtype)
 
     def get_queryset(self):
-        return self.get_model().objects.all()
\ No newline at end of file
+        return self.get_model().objects.all()
+    
+class AdDetailView(LoginRequiredMixin, DetailView, ModelMappingMixin):
+    template_name = 'ad/ad_detail.html'
+    context_object_name = 'ad'
+
+    def get_model(self):
+        adtype = self.kwargs['adtype']
+        return self.model_mapping.get(adtype)
```

Comment:
Add testcase for a case if adtype value is invalid and it does not have any model mapping.